data/routesPosts.js

```diff
@@ -0,0 +1,100 @@
+const express = require('express');
+
+const PostDB = require('./helpers/postDb.js');
+
+const routesPosts = express.Router();
+
+
+// Server.js handles the /api path. Router.js handles the posts path
+
+
+// Get all the posts
+routesPosts.get('/', async (req, res) => {
+    try {
+        const posts = await PostDB.get(req.query);
+        res.status(200).json(posts);
+    } catch (error) {
+        res.status(500).json({ error: "The posts could not be retrieved." });
+    }
+});
+
+
+// Get single post by id
+routesPosts.get('/:id', async (req, res) => {
+    try {
+        const postById = await PostDB.getById(req.params.id);
+        if (postById.length === 0) {  
+            res.status(404).json({ message: "The post with the specified ID does not exist." });
+        } else {
+            res.status(200).json(postById);
+        }
+    } 
+    catch (error) { 
+        res.status(500).json({ error: "The post could not be retrieved." });
```

Comment:
hmm, there's something fishy going on with this endpoint and it might be stemming from your conditional statement--when I attempt to access posts/111 (non-existent) it returns an error instead of the desired 404. That is probably because postById will not be an object or array that you can check the length of if it doesn't exist, so postById.length can never equal zero. Instead check the inherent truthiness of the postById coming back from that promise instead, if it exists it will be truthy, if it is undefined (which I think it might be, check via console.log or a logging middleware) it will be falsey

Reply:
Ahhh, excellent catch. 

Fixed by checking for the existence of !postById. If that is truthy, returns the 404. Otherwise, returns the 200.